packages/plugin-hardhat/src/utils/deploy-impl.ts

```diff
@@ -14,53 +14,152 @@ import {
   inferProxyKind,
   setProxyKind,
   ValidationOptions,
+  getBeaconAddress,
 } from '@openzeppelin/upgrades-core';
 
 import { deploy } from './deploy';
 import { Options, withDefaults } from './options';
 import { readValidations } from './validations';
+import { getIBeaconFactory } from '.';
+import { FormatTypes } from 'ethers/lib/utils';
 
-interface DeployedImpl {
+interface DeployedProxyImpl {
   impl: string;
   kind: NonNullable<ValidationOptions['kind']>;
 }
 
-export async function deployImpl(
+interface DeployedBeaconImpl {
+  impl: string;
+}
+
+export enum DeployType {
+  Proxy,
+  Beacon,
+}
+
+export async function deployProxyImpl(
+  hre: HardhatRuntimeEnvironment,
+  ImplFactory: ContractFactory,
+  opts: Options,
+  proxyOrBeaconAddress?: string,
+): Promise<DeployedProxyImpl> {
+  return deployImpl(hre, ImplFactory, opts, false, proxyOrBeaconAddress);
+}
+
+export async function deployBeaconImpl(
+  hre: HardhatRuntimeEnvironment,
+  ImplFactory: ContractFactory,
+  opts: Options,
+  proxyOrBeaconAddress?: string,
+): Promise<DeployedBeaconImpl> {
+  return deployImpl(hre, ImplFactory, opts, true, proxyOrBeaconAddress);
+}
+
+async function deployImpl(
   hre: HardhatRuntimeEnvironment,
   ImplFactory: ContractFactory,
   opts: Options,
-  proxyAddress?: string,
-): Promise<DeployedImpl> {
+  isBeacon: boolean,
+  proxyOrBeaconAddress?: string,
```

Comment:
I realize now that we can do something simpler and more focused in `deployImpl`, and I think we should.

It seems the only thing that `isBeacon` affects is how we retrieve the implementation address. So my proposal is to move that logic outside of here, and instead of receiving the address of a proxy or a beacon as an argument, receive the implementation address directly. Doing that, `deployImpl` should be a lot simpler. The `kind` return argument can be ignored for beacon proxies.

Let me know what you think!

Reply:
I agree.  This would allow `deployImpl` to just focus on the core logic of validation and deployment.  The scenario-specific parts could perhaps be moved to the `deployProxyImpl` and `deployBeaconImpl` functions above.